Config changes per Issue #11

dgou opened the pull request:
@peterkaminski 
Submitting this as a pull request, but still WIP.
All the config file loading stuff is done.
The copy/diff part is _NOT_ done because I have questions about how backups work when we start cherry-picking files deeper in the vault configuration than the top level directories.

To requests:
1. Review the code so far.  README is yet to come!
2. Discuss in https://github.com/peterkaminski/obsidian-settings-manager/issues/11 what we want the backup strategy to look like.

dgou committed:
Fix the JSON loading helper function name.

Keeping the next diff cleaner.

dgou committed:
Convert to JSON for configuration, add command line switch.

Next commits will implement based on comments/discussion from:
https://github.com/peterkaminski/obsidian-settings-manager/issues/11

dgou committed:
Load our own (OSM) config file or use internal default contents

dgou committed:
Fix bug with quoting strings.

The encoded JSON needs to be a raw string to prevent backslash quoting explosion

dgou committed:
Self-Documentation for future readers

dgou committed:
Document code sections and move code around.

No code changes, just adding comments and moving code.
The diff is bad enough without having to look for a code change too.

dgou committed:
Load the OSM config file.

Added primitive tracing to help when things do work as expected.
Very preliminary implmentation. ENV VARs are a quick way to
set values before the command line is parsed but aren't as user friendly.

dgou committed:
Be persnickety about the structure of the config files.

dgou committed:
Look for the Obsidian config file and load it.

dgou committed:
Load the Obsidian Vault paths.

This also involved tracking the Obsidian Root dirctory since it is
no longer a parameter or env var to the script

dgou committed:
Add option to print the default configuration.

Also, hoist the code for version and config printing to above
where the configs are loaded. They don't depend on it, and this makes
it easier to bootstrap a custom configuration.

dgou committed:
Allow the obsidian config file override to use the `~` short hand.

dgou committed:
Misc fixes: Actually use value of --config parameter, trace where obsidian config is loading from.

dgou commented on osm.py:
```diff
@@ -401,6 +417,8 @@ def main():
             print("Error: Cannot locate the 'diff' command, aborting.")
             exit(-1)
 
+    load_osm_config()
```
Yeah, I know, I forgot the args.config parameter here, it will be fixed in a later commit :-)

dgou committed:
New helper for checking types in the config files.

Helps us to be more careful about using the config.
Usages to follow in the next commits.

dgou committed:
Parse the files_to_copy list into a sequence of actions.

Next commits will implement finding the files in a specific vault.

dgou commented on osm.py:
```diff
@@ -492,6 +538,7 @@ def main():
 
     try:
         vault_paths = get_vault_paths()
+        directives = list(get_processing_directives())
```
I was on the fence about where in the code to do this as it is a defacto check on the configuration file even if no file operations are going to be performed.
On the one hand, validating early is good.
On the the other hand, validating when it won't be used (say for --list) could make it harder for someone to incrementally create/tweak their config...

peterkaminski replied:
Either way is fine, of course - I can make UX arguments either way, and so for me, the UX arguments are kind of a wash.

Tiebreaker might be simplicity / readability of the code itself, and I think that argues for validating early.

dgou replied:
Yeah, I think I am ok with early too until a compelling argument comes in for postponing.

dgou committed:
Wordsmithing on the bad actions message.

Probably needs more work too.

dgou committed:
Implement the file action matching functions.

dgou committed:
Keep files to be copied sorted.

It's much easier on the human brain to see them in sorted order
esp when looking at diffs.

dgou committed:
Option to show what files would be selected for a particular vault.

Sure, you could use `--diff-to` and try to sort it out, but as you
are adjusting your files_to_copy list, it would be nice to see
how it would pull files from a specific vault without the other output

dgou commented:
Putting pen down as of commit 19 to sort out the backups thing.

dgou committed:
Fix baackup action function name.

Was left over from when there were two separate backup action functions.

dgou committed:
Add configuration info for backups.

dgou commented on osm.py:
```diff
@@ -106,7 +106,11 @@
-    ]
+    ],
+    "backup": {
+        "archive_format_priority_list": ["zip", "gztar", "bztar", "xztar", "tar"],
```
@peterkaminski the idea here is that since any one format might not be available, better to have a list of preferred formats, and work left to right until `make_archive` works.
I'm open to better names for the keys/values here.

peterkaminski replied:
looks fine to me.

dgou committed:
Remove old per-file backup functionality.

dgou commented:
Whoa, has it been two weeks!  Crikey!
Planning on working more on this today, day-job has been unexpectedly long hours.

peterkaminski commented:
> Whoa, has it been two weeks! Crikey! Planning on working more on this today, day-job has been unexpectedly long hours.

Me too! 🙂

dgou committed:
Allow iso date format to be more generally used.

Remove the leading dash so that it can be used as a prefix.
Match the ISO string anywhere, prefix, suffix, or embedded.

dgou committed:
Add new backup commands and implement backup_vault code.

Also backup before doing an update.

NOTE: update hasn't been fixed yet,
      only the backup code is working at this point

dgou commented:
@peterkaminski - take a look at the latest two commits.
the update functionality isn't done yet, but I wanted to get your thoughts on the new backup code, the new command options, etc.

dgou commented on osm.py:
```diff
@@ -531,11 +535,42 @@ def diff_settings(dest, src):
+        try:
+            backup_file = shutil.make_archive(backup_file, backup_format, root_dir=backup_from_dir)
+            if backup_file:
+                print("Backup made:", backup_file)
```
If we like how this works, I'll update to have versbose and dry-run implemented. The trick with dry-run though is that until we try to call `make_archive` we don't know for sure sure sure if that format is actually supported.
There is https://docs.python.org/3/library/shutil.html#shutil.get_archive_formats but I am not sure if those are guaranteed to work, or if they are just potentially available...

peterkaminski replied:
`shutil.make_archive()` takes a `dry_run` boolean. ;-)

dgou replied:
D'oh!  So it does! I think I knew that when I first found it, but then forgot, thanks!

peterkaminski commented:
> the update functionality isn't done yet, but I wanted to get your thoughts on the new backup code, the new command options, etc.

@dgou, looks good to me.

dgou committed:
Implment verbose and dry-run for backing up vaults

dgou commented on osm.py:
```diff
@@ -536,7 +536,7 @@ def diff_settings(dest, src):
 
 def backup_vault(vault_path):
-    print('#', vault_path)
+    print(f'# Backing up settings for: {vault_path}')
```
I wasn't sure about keeping this as `print` vs `verbose` but when I was testing it I liked seeing that it was doing something and where, but I'm open to making this `verbose` instead.
